crashtracker/src/crash_handler.rs

```diff
@@ -255,9 +295,30 @@ fn handle_posix_signal_impl(signum: i32) -> anyhow::Result<()> {
     Ok(())
 }
 
-// TODO, there is a small race condition here, but we can keep it small
+/// Registers UNIX signal handlers to detect program crashes.
+/// This function can be called multiple times and will be idempotent: it will
+/// only create and set the handlers once.
+/// However, note the restriction below:
+/// PRECONDITIONS:
+///     The signal handlers should be restored before removing the reciever.
+/// SAFETY:
+///     Crash-tracking functions are not guaranteed to be reentrant.
+///     No other crash-handler functions should be called concurrently.
+/// ATOMICITY:
+///     This function uses a compare_and_exchange on an atomic pointer.
+///     However, setting the crash handler itself is not an atomic operation
+///     and hence it is possible that a concurrent operation could see partial
+///     execution of this function.
+///     If a crash occurs during execution of this function, it is possible that
+///     the crash handler will have been registered, but the old signal handler
+///     will not yet be stored.  This would lead to unexpected behaviour for the
+///     user.  This should only matter if something crashes concurrently with
+///     this function executing.
 pub fn register_crash_handlers(create_alt_stack: bool) -> anyhow::Result<()> {
-    anyhow::ensure!(OLD_HANDLERS.load(SeqCst).is_null());
+    if !OLD_HANDLERS.load(SeqCst).is_null() {
+        return Ok(());
+    }
```

Comment:
Minor: Not sure this is something we want to care about, but one thing we could do when the function is called is to check that our signal handlers are still active, e.g. because something else on the system may have overwritten them.

(I'm thinking of this, because there may be a situation where someone starts the crash tracker, and then discovers that the VM overwrites their signal handlers after X, and then calls `register_crash_handlers` again expecting libdatadog to be restored but isn't. Maybe too farfetched?)

Reply:
https://github.com/DataDog/libdatadog/issues/325